_posts/2020-07-29-parallelized-sampling/parallelized-sampling.Rmd

```diff
@@ -0,0 +1,139 @@
+- Bonus: it is also suitable for weighted reservoir sampling (i.e., can sample $n$ out of a possibly infinite stream of rows according to their weights such that at any moment the $n$ samples will be a weighted representation of all rows that have been processed so far).
+
+## Why does this algorithm work?
+
+As an interesting aside, some readers have probably seen this technique presented in a slightly different form under another name. It is in fact equivalent to a generalized version of the [Gumbel-max trick](https://lips.cs.princeton.edu/the-gumbel-max-trick-for-discrete-distributions) which is commonly referred to as the Gumbel-top-k trick. Readers familiar with properties of the Gumbel distribution will no doubt have an easy time convincing themselves the algorithm above works as expected.
+
+In this section, I will also present a proof of correctness for this algorithm based on elementary properties of [probability density function](https://en.wikipedia.org/wiki/Probability_density_function) (shortened as PDF from now on), [cumulative distribution function](https://en.wikipedia.org/wiki/Cumulative_distribution_function) (shortened as CDF from now on), and basic calculus.
+
+First of all, to make sense of all the $\ln(u_j) / w_j$ calculations in this algorithm, one has to understand [inverse transform sampling](https://en.wikipedia.org/wiki/Inverse_transform_sampling). For each $j \in \{1, \dotsc, N\}$, consider the probability distribution defined on $(-\infty, 0)$ with CDF $F_j(x) = e^{w_j \cdot x}$. In order to pluck out a value $y$ from this distribution, we first sample a value $u_j$ uniformly randomly out of $(0, 1)$ that determines the percentile of $y$ (i.e., how our $y$ value ranks relative to all possible $y$ values, a.k.a, the "overall population", from this distribution), and then apply $F_j^{-1}$ to $u_j$ to find $y$, so, $y = F_j^{-1}(u_j) = \ln(u_j) / w_j$.
+
+Secondly, after defining all the required CDF functions $F_j(x) = e^{w_j \cdot x}$ for $j \in \{1, \dotsc, N\}$, we can also easily derive their corresponding PDF functions $f_j$: $$f_j(x) = \frac{d F_j(x)}{dx} = w_j e^{w_j \cdot x}$$.
+
+Finally, with a clear understanding of the family of proability distributions this algorithm is working with, it is a non-difficult calculus exercise to show the probability of selecting any tuple of rows (say $(r_1, \dotsc, r_n)$) will be $\prod\limits_{j = 1}^{n} \left( {w_j} \middle/ {\sum\limits_{k = j}^{N}{w_k}} \right)$, which implies the possible outcomes of this algorithm will have exactly the same probability distribution as that of a $n$-step `SWoR`.
+
+In order to not deprive our dear readers the pleasure of completing this proof by themselves, I decided not to inline the rest of the proof within this blog post, and make everything available in [this file](proof.pdf) instead.
+
```

Comment:
I'd add add the code they need to try this out in `sparklyr`, maybe with something like:

````
This new algorithm just got implemented into `sparklyr` with PR #123. To try it out, install from GitHub followed by using `sample_frac` with the `weight` parameter:

```
remotes::install_github("sparklyr/sparklyr")

library(sparklyr)
sc <- spark_connect(master = "local")

copy_to(sc, mtcars) %>% 
  dplyr::sample_frac(0.25, weight = gear)
```
```
# ... output ...
```
````